### PR TITLE
[stdlib] Add `Writable.estimate_bytes_to_write()` with a default implementation

### PR DESCRIPTION
**Pull request by martinvuyk**
Part of #4249. This will enable every `Writable` type to implement a faster size estimation function in cases where calling `self.write_to(writer)` is expensive. It will also allow other `Writer` types like string to previously reserve the necessary buffer size, enabling equivalent optimizations to what was done by using the `_TotalWritableBytes` internal tool.

This should eventually be extended to enable one to pass the format specification and calculate the necessary buffer size according to it as well.

This also adds integer write size estimation size for `SIMD` as a proof of the improvements this enables

## Benchmark results

CPU: Intel® Core™ i7-7700HQ

|name | old value (ms) | new value (ms) | markdown perc | magnitude|
|-- | -- | -- | -- | -- |
|bench_string_write_int_short | 17.0325 | 17.2421 | -0.0123 | 0.9878|
|bench_string_write_int_long | 37.1422 | 0.0881 | 0.9976 | 421.5114|

### CODE REVIEW

File: mojo/stdlib/std/format/__init__.mojo [hunk 1]
```diff
@@ -251,6 +251,22 @@ trait Writable:
 
         _reflection_write_to[f=call_write_repr_to](self, writer)
 
+    @always_inline("nodebug")
+    fn estimate_bytes_to_write(self) -> UInt:
```
Comment (NathanSWard) [inline mojo/stdlib/std/format/__init__.mojo]:
```suggestion
    fn estimate_bytes_to_write(self) -> Int:
```
We're standardizing on `Int` being the canonical return type in the stdlib :)

File: mojo/stdlib/std/format/__init__.mojo [hunk 2]
```diff
@@ -251,6 +251,22 @@ trait Writable:
 
         _reflection_write_to[f=call_write_repr_to](self, writer)
 
+    @always_inline("nodebug")
```
Comment (NathanSWard) [inline mojo/stdlib/std/format/__init__.mojo]:
```suggestion
    @always_inline
```
I definitely believe we can remove the `"nodebug"` - not too sure what benefit we get by hiding the debug info for this function.

Also not entirely sure of the benefit of `always_inline` here either, it's probably okay but I wonder if there's actually a benefit. Especially since this may greatly increase binary size if the `write_to` of `self` is massive and inlined as well.

File: mojo/stdlib/std/format/_utils.mojo [hunk 3]
```diff
@@ -392,6 +392,7 @@ struct _TotalWritableBytes(Writer):
             for i in range(1, length):
                 self.write(sep, values[i])
 
+    @always_inline("nodebug")
```
Comment (NathanSWard) [inline mojo/stdlib/std/format/_utils.mojo]:
```suggestion
    @always_inline
```

File: mojo/stdlib/std/collections/string/string.mojo [hunk 4]
```diff
@@ -645,23 +645,22 @@ struct String(
             args: Sequence of arguments to write to this Writer.
         """
         comptime length = args.__len__()
-        var total_bytes = _TotalWritableBytes()
-        total_bytes.size += self.byte_length()
+        var total_bytes = UInt(0)
```
Comment (NathanSWard) [inline mojo/stdlib/std/collections/string/string.mojo]:
```suggestion
        var total_bytes = self.byte_length()
```

The previous implementation had the initial byte_length of the string. Should we keep this?

File: mojo/stdlib/std/collections/string/string.mojo [hunk 5]
```diff
@@ -645,23 +645,22 @@ struct String(
...
             @parameter
             for i in range(length):
                 args[i].write_to(writer)
 
-        if total_bytes.size <= Self.INLINE_CAPACITY:
+        if Int(total_bytes) <= Self.INLINE_CAPACITY:
             _write(self)
         else:
-            self.reserve(total_bytes.size)
+            self.reserve(Int(total_bytes))
```
Comment (NathanSWard) [inline mojo/stdlib/std/collections/string/string.mojo]: One potential issue I see with this current design is what if the `estimate_bytes_to_write` ends up overshooting the total number of bytes needed. So we are left with an excess heap allocation.

It might be worth adding in the docs of `estimate_bytes_to_write` that this should be either an exact or conservative estimate.

File: mojo/stdlib/std/format/__init__.mojo [hunk 6]
```diff
@@ -251,6 +251,22 @@ trait Writable:
...
+            The estimated bytes needed to write the type into a writer.
+
+        Notes:
+            By default this function runs the type's `write_to` function. If
+            doing that is not cheap for a given type, consider writing a faster
+            size estimation function instead.
+        """
+        var total_bytes = _TotalWritableBytes()
+        self.write_to(total_bytes)
+        return UInt(total_bytes.size)
```
Comment (NathanSWard) [inline mojo/stdlib/std/format/__init__.mojo]: Can we add a unit test in `test/format` that binds this default behavior?

**Review by NathanSWard — CHANGES_REQUESTED** [review]

File: mojo/stdlib/std/format/__init__.mojo (hunk 2 above)
Comment (martinvuyk) [reply mojo/stdlib/std/format/__init__.mojo]: I think that this should all inline into just adding integers, otherwise we have massive call chains for very trivial ops. And having this in debug will just add noise for a single add operation and each call to estimate size. The whole idea about this is to simplify a lot of the function calls just down to arithmetic

File: mojo/stdlib/std/collections/string/string.mojo (hunk 4 above)
Comment (martinvuyk) [reply mojo/stdlib/std/collections/string/string.mojo]: Yeah I missed this while copy pasting code, will fix

File: mojo/stdlib/std/format/__init__.mojo (hunk 2 above)
Comment (NathanSWard) [reply mojo/stdlib/std/format/__init__.mojo]: It's fine to keep it always_inline - we just need to be very conscious about these are we are way to liberal with it across the codebase generally.

I'd still request you remove `"nodebug"` please.

**Comment by martinvuyk** [comment]
@NathanSWard added code for `SIMD` integers and benchmarks to prove the improvement this new setup enables. I've updated the PR description with the results 421x improvement (for sequential writing) :tada: :rocket:

File: mojo/stdlib/std/builtin/simd.mojo [hunk 7]
```diff
@@ -2282,6 +2284,49 @@ struct SIMD[dtype: DType, size: Int](
             _write_scalar(writer, element)
         writer.write_string(")")
 
+    fn estimate_bytes_to_write(self) -> Int:
```
Comment (NathanSWard) [inline mojo/stdlib/std/builtin/simd.mojo:2276]: I would recommend pushing off the `SIMD.estimate_bytes_to_write` to a separate PR - or if you wanted to add it in this PR, we now need to add tests for this function as well :)

File: mojo/stdlib/std/math/math.mojo [hunk 8]
```diff
@@ -2223,10 +2223,10 @@ fn log10[dtype: DType, width: Int, //](x: SIMD[dtype, width]) -> type_of(x):
     __comptime_assert (
         dtype.is_floating_point()
     ), "input type must be floating point"
+    comptime `log10(2)` = 0.301029995663981195213738894724493027
 
     @parameter
     if is_nvidia_gpu():
-        comptime log10_2 = 0.301029995663981195213738894724493027
```
Comment (NathanSWard) [inline mojo/stdlib/std/math/math.mojo]: This looks like a drive-by change, but I'm not a fan of using the back-tick variable name here, IMO it a bit messier than before. `log10_2` was a okay name.

Comment (soraros) [reply mojo/stdlib/std/math/math.mojo]: You could add a inline `llvm_intrinsic["llvm.log10", ...](...)`: https://github.com/modular/modular/issues/5856#issuecomment-3802789632

File: mojo/stdlib/std/builtin/simd.mojo (hunk 7 above)
Comment (martinvuyk) [reply mojo/stdlib/std/builtin/simd.mojo:2276]: I wanted to have some proof of the improvements this allows. If you want I can split it off, the implementation got a bit more complicated than I originally expected. I did add tests that cover pretty much every edge case now.

File: mojo/stdlib/std/builtin/simd.mojo [hunk 9]
```diff
@@ -2284,6 +2287,120 @@ struct SIMD[dtype: DType, size: Int](
...
+        """Estimate the total bytes to write the type into a writer.
+
+        Returns:
+            The estimated bytes needed to write the type into a writer.
+        """
+
+        @parameter
+        if Self.dtype.is_integral():
```
Comment (soraros) [inline mojo/stdlib/std/builtin/simd.mojo]: Nit: could you swap the two branches, so the simpler case (`not is_integral()`) is handled first? I think that'd make the code a bit more readable.

File: mojo/stdlib/std/builtin/simd.mojo [hunk 10]
```diff
@@ -2284,6 +2287,120 @@ struct SIMD[dtype: DType, size: Int](
...
+                # 4. After this comparison 32bit integers need only 8 digit
+                # comparisons instead of 16 and 64bit need only 16 instead of 32
+                var comp: Bool
+
+                @parameter
+                if is_unsigned:
+                    comp = self < 10_000
+                else:
+                    comp = Self(-10_000) < self < Self(10_000)
+                if comp:
```
Comment (soraros) [inline mojo/stdlib/std/builtin/simd.mojo]: Nit: could you change this part to the following structure:
```mojo
if not comp:
  return ...
longer_body
return ...
```

**Comment by NathanSWard** [comment]
Awesome, this is looking good! I'll get to this first thing next week and get it synced in :)

**Comment by NathanSWard** [comment]
Hey @martinvuyk - if you can rebase, I'll get this merged in asap.
I can already think of places to apply it in the updated formatting code :)

File: mojo/stdlib/std/builtin/variadics.mojo [hunk 11]
```diff
@@ -1040,6 +1040,42 @@ struct VariadicPack[
                 self[i].write_to(writer)
         writer.write_string(end)
 
+    fn _estimate_bytes_to_write[
```
Comment (martinvuyk) [inline mojo/stdlib/std/builtin/variadics.mojo:1043]: @NathanSWard I just added this as a fly-by change to cleanup string's use of it. I can split it off if you wish

Comment (martinvuyk) [reply mojo/stdlib/std/builtin/variadics.mojo:1043]: The amount of repetition when doing this for collections is tickling at my brain for a shared implementation. I don't know if we'll ever have something like `ComptimeIterable` or `ComptimeIndexable` + the equivalent runtime things and implement this and write_to generically for iterable collections

Comment (NathanSWard) [reply mojo/stdlib/std/builtin/variadics.mojo:1043]: We ideally want to be generic over `Iterable`, however our current iterator traits don't let you return references in generic contexts, meaning if we had a `write_iterable_to`, it would invoke a copy of every element unfortunately. 

It'll first take some reworking of `Iterable` before we can do this generically unfortunately.

Comment (NathanSWard) [reply mojo/stdlib/std/builtin/variadics.mojo:1043]: We hopefully wouldn't need to do "comptime iterable" as the `Iterable` trait should allows ways to do this.
Things like defining an optional static size, or something like Rust's `ExactSizeIterator` are possible routes.

**Comment by NathanSWard** [comment]
!sync

File: mojo/stdlib/std/builtin/simd.mojo [hunk 12]
```diff
@@ -2271,6 +2273,119 @@ struct SIMD[dtype: DType, size: Int](
             _write_scalar(writer, element)
         writer.write_string(")")
 
+    fn estimate_bytes_to_write(self) -> Int:
```
Comment (NathanSWard) [inline mojo/stdlib/std/builtin/simd.mojo:2276]: @martinvuyk - do you have a reference/link to the algorithm used here? E.g. if it's a paper or if it's implemented in another language?

**Review by NathanSWard — APPROVED** [review]

File: mojo/stdlib/std/builtin/simd.mojo (hunk 12 above)
Comment (martinvuyk) [reply mojo/stdlib/std/builtin/simd.mojo:2276]: Reference: my braincells lol

I like inventing the stuff I use, it's what I like about contributing to Mojo; I get to try and come up with something no one did before and try to beat every other implementation

Comment (NathanSWard) [reply mojo/stdlib/std/builtin/simd.mojo:2276]: Cool, that works too. Just curious to if there is a reference implementation, as it's easier to debug/maintain especially if we end up finding bugs in the future.

**Comment by NathanSWard** [comment]
Hey @martinvuyk!
Talked with the stdlib team. We're in a phase where we want to be judicious and about what new APIs we add before 1.0, since every API added now is one we'll have to review as a potential candidate to stabilize. Given that this feature is not found in many other standard libraries, and there are available workaround like manual pre-allocation, I suggest we back-burner this until later when we can give it a more careful consideration?

A few questions/concerns about this:
1. With the current API, if an implementor of this function _under_ estimates the size, we have the issues where the backing string would need to grow (reallocate) once the values are written to, which slightly defeats the purpose of this utility.
- I could see a use for `get_exact_size` or something like that which uses the `_WritableBytes` helper type under the hood if the user _needs_ the exact number of bytes for preallocation (where either under or over allocating would not be acceptable).
2. Does this utility have an application outside of pre-allocating for `String`? 
- For example, if we write directly to an underlying buffer (File handle) we don't need the check the size before hand.
- The `Writable` trait is really only useful when used for debugging, I could see some use in a more generic (de)serialization library, however I'm unsure if this optimization will really be necessary in generic debug/printing code.

**Comment by martinvuyk** [comment]
> Talked with the stdlib team. We're in a phase where we want to be judicious and about what new APIs we add before 1.0, since every API added now is one we'll have to review as a potential candidate to stabilize

@NathanSWard I completely understand and agree, this should not be stabilized yet. FYI I also don't want the separation of `write_to` from `write_to_repr` to be stabilized yet, `Writable` is not only about debugging and that separation is already giving issues for nested structures (so even debugging is kind of harder/uglier). 

> Given that this feature is not found in many other standard libraries, and there are available workaround like manual pre-allocation

I'm honestly not surprised given the only language where I've liked dealing with strings is Python, and my goal is to make Mojo's even better.

> I suggest we back-burner this until later when we can give it a more careful consideration?

Yeah I think it's gotten to a point where I should write out a proposal to explain a bit more of my motivations and vision for the whole `Writable` and `Writer` APIs. I'll be writing it in my free time this week and open a PR shortly.

> 1. With the current API, if an implementor of this function _under_ estimates the size, we have the issues where the backing string would need to grow (reallocate) once the values are written to, which slightly defeats the purpose of this utility.
> 
> 
>     * I could see a use for `get_exact_size` or something like that which uses the `_WritableBytes` helper type under the hood if the user _needs_ the exact number of bytes for preallocation (where either under or over allocating would not be acceptable).

I think we could adapt this by having the allowance/preference of over or under-estimation be determined by an argument. It does deserve a bit more consideration, and we could start by making it an exact requirement and thus the fallback to `_TotalWritableBytes` and later on relax it to include e.g. float estimation (which is over-estimating in the algorithm I've managed to setup) which is another case where transforming it to a string is very expensive

> 2. Does this utility have an application outside of pre-allocating for `String`?

Anything that has to do with manipulating strings, and potentially some other logic like IO could make use of the knowledge of how much data is to be transmitted in total.
 
>     * For example, if we write directly to an underlying buffer (File handle) we don't need the check the size before hand.

If you want to assume the file system has enough space for storing the string sure, but if you're working on constrained environments where e.g. writing out an array of numbers takes more memory than what is available either in the file system or in RAM, then we need to have this kind of setup.

>     * The `Writable` trait is really only useful when used for debugging, I could see some use in a more generic (de)serialization library, however I'm unsure if this optimization will really be necessary in generic debug/printing code.

I will show more use-cases in the proposal I'll write, but suffice it to say that debugging is just the tip of the iceberg in what I want to achieve with `Writable` and `Writer`. If we want to follow Python's lead (which I strongly want to since I like their string abilities), we need to think more in the direction of template strings. I want our trait family to be used for everything from json, yaml, to html and beyond. I think we have a huge opportunity here to setup something that is easily extensible by external libraries in such a way that we could provide native-feeling templating infrastructure
